File: python/test/t_EnclosingSimplexMonotonic1D_std.py
#! /usr/bin/env python

from __future__ import print_function
import openturns as ot
import os

ot.TESTPREAMBLE()

nrVertices = 100
vertices = ot.Normal().getSample(nrVertices).sort()
simplices = [[i, i + 1] for i in range(nrVertices - 1)]
mesh1 = ot.Mesh(vertices, simplices)
vertices *= -1.0
mesh2 = ot.Mesh(vertices, simplices)

for mesh in [mesh1, mesh2]:
    lowerBound = mesh.getLowerBound()[0]
    upperBound = mesh.getUpperBound()[0]
    n = mesh.getSimplicesNumber()
    print("mesh=", mesh, "lowerBound=", lowerBound,
          "upperBound=", upperBound, n, "simplices")
    algo = ot.EnclosingSimplexMonotonic1D(mesh.getVertices())

    ot.RandomGenerator.SetSeed(0)
    test = ot.Sample(ot.Uniform(-3.0, 3.0).getSample(1000))

    vertices = mesh.getVertices()
    for vertex in test:
        index = algo.query(vertex)
        x = vertex[0]
        if x < lowerBound or x > upperBound:
            if index < n:
                print(
                    "Point", x, "should be outside but query returned index", index)
                os.exit(1)
        else:
            if index >= n:
                print(
                    "Point", x, "should be inside but query returned index", index)
                os.exit(1)

            found, coordinates = mesh.checkPointInSimplexWithCoordinates(
                vertex, index)
            if not found:
                print("Wrong simplex found for", x,
                      "(index=", index, ") barycentric coordinates=", coordinates)
                os.exit(1)
            if coordinates[0] < 0.0 or coordinates[0] > 1.0 or coordinates[1] < 0.0 or coordinates[1] > 1.0:
                print("Wrong barycentric coordinates found found for", x,
                      "(index=", index, ") barycentric coordinates=", coordinates)
                os.exit(1)
            if abs(x - coordinates[0] * vertices[index, 0] - coordinates[1] * vertices[index + 1, 0]) > 1.e-10:
                print("Wrong barycentric coordinates found found for", x,
                      "(index=", index, ") barycentric coordinates=", coordinates)
                os.exit(1)

File: lib/test/t_EnclosingSimplexMonotonic1D_std.cxx
//                                               -*- C++ -*-
/**
 *  @brief The test file of class EnclosingSimplexMonotonic1D for standard methods
 *
 *  Copyright 2005-2022 Airbus-EDF-IMACS-ONERA-Phimeca
 *
 *  This library is free software: you can redistribute it and/or modify
 *  it under the terms of the GNU Lesser General Public License as published by
 *  the Free Software Foundation, either version 3 of the License, or
 *  (at your option) any later version.
 *
 *  This library is distributed in the hope that it will be useful,
 *  but WITHOUT ANY WARRANTY; without even the implied warranty of
 *  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 *  GNU Lesser General Public License for more details.
 *
 *  You should have received a copy of the GNU Lesser General Public License
 *  along with this library.  If not, see <http://www.gnu.org/licenses/>.
 *
 */
#include "openturns/OT.hxx"
#include "openturns/OTtestcode.hxx"

using namespace OT;
using namespace OT::Test;

int main(int, char *[])
{
  TESTPREAMBLE;
  OStream fullprint(std::cout);

  const UnsignedInteger nrVertices = 100;
  Sample vertices(Normal().getSample(nrVertices).sort());
  IndicesCollection simplices(nrVertices - 1,  2);
  for(UnsignedInteger i = 0;  i < nrVertices - 1; ++i)
  {
    simplices(i, 0) = i;
    simplices(i, 1) = i + 1;
  }
  Collection<Mesh> meshes1D;
  meshes1D.add(Mesh(vertices, simplices));
  vertices *= -1.0;
  meshes1D.add(Mesh(vertices, simplices));
  for(UnsignedInteger m = 0; m < meshes1D.getSize(); ++m)
  {
    const Mesh mesh = meshes1D[m];
    const Scalar lowerBound = mesh.getLowerBound()[0];
    const Scalar upperBound = mesh.getUpperBound()[0];
    const UnsignedInteger n  = mesh.getVerticesNumber() - 1;
    fullprint << "mesh = " << mesh << std::endl;
    const Sample meshVertices(mesh.getVertices());
    const EnclosingSimplexMonotonic1D algo(meshVertices);

    RandomGenerator::SetSeed(0);
    const Sample test(Uniform(-3.0, 3.0).getSample(1000));

    Point coordinates;
    for (UnsignedInteger i = 0; i < test.getSize(); ++i)
    {
      UnsignedInteger index = algo.query(test[i]);
      if (test(i, 0) < lowerBound || test(i, 0) > upperBound)
      {
        if (index < n)
        {
          fullprint << "Point " << test[i] << " should be outside but query returned index " << index << std::endl;
          return ExitCode::Error;
        }
      }
      else
      {
        if (index >= n)
        {
          fullprint << "Point " << test[i] << " should be inside, query returned index " << index << std::endl;
          return ExitCode::Error;
        }
        if (!mesh.checkPointInSimplexWithCoordinates(test[i], index, coordinates))
        {
          fullprint << "Wrong simplex found for " << test[i] << " (index=" << index << ") barycentric coordinates=" << coordinates << std::endl;
          return ExitCode::Error;
        }
        if (coordinates[0] < 0.0 || coordinates[0] > 1.0 || coordinates[1] < 0.0 || coordinates[1] > 1.0)
        {
          fullprint << "Wrong barycentric coordinates found found for " << test[i] << " (index=" << index << ") barycentric coordinates=" << coordinates << std::endl;
          return ExitCode::Error;
        }
        if (std::abs(test(i, 0) - coordinates[0] * meshVertices(index, 0) - coordinates[1] * meshVertices(index + 1, 0)) > 1.e-10)
        {
          fullprint << "Wrong barycentric coordinates found found for " << test[i] << " (index=" << index << ") barycentric coordinates=" << coordinates << std::endl;
          return ExitCode::Error;
        }
      }
    }
  }
  return ExitCode::Success;
}
